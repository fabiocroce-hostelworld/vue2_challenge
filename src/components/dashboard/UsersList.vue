<template>
  <v-card>
    <v-card-title> Users </v-card-title>
    <v-card-text>
      <v-list>
        <v-list-item-group v-model="selectedUsers.users" multiple @change="handleChange(selectedUsers)">
          <template v-for="(item, i) in users">
            <v-list-item :key="`item-${i}`" :value="item.userId" active-class="primary--text">
              <template v-slot:default="{ active }">
                <v-list-item-content>
                  <v-list-item-title v-text="item.user"></v-list-item-title>
                </v-list-item-content>

                <v-list-item-action>
                  <v-checkbox :input-value="active" color="primary" class="user-checkbox" />
                </v-list-item-action>
              </template>
            </v-list-item>
          </template>
        </v-list-item-group>
        <v-divider class="mb-4" />
        <v-btn class="toggle-all mr-4 primary" @click="toggleAll(users, selectedUsers)">
          Toggle all shifts
        </v-btn>
      </v-list>
    </v-card-text>
  </v-card>
</template>

<script lang="ts">
import Vue from "vue"
import { reactive } from "@vue/composition-api"
import { User } from "@/models"

type Users = Array<User>
type SelectedUsers = { users: Array<number> }

export default Vue.extend({
  name: "UsersList",
  props: {
    users: {
      type: Array,
      required: true,
      validator: (prop: Users) => prop.every((user: User) => user instanceof User)
    }
  },
  setup(_, { emit }) {
    const selectedUsers = reactive<SelectedUsers>({ users: [] })

    const handleChange = (currentSelectedUsers: SelectedUsers) => {
      emit("change", currentSelectedUsers.users)
    }

    const toggleAll = (users: Users, currentSelectedUsers: SelectedUsers) => {
      currentSelectedUsers.users = currentSelectedUsers.users.length ? [] : users.map((user: User) => user.userId)
      emit("change", currentSelectedUsers.users)
    }

    return {
      selectedUsers,
      handleChange,
      toggleAll
    }
  }
})
</script>
