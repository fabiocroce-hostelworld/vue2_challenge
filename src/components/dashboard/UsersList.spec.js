import { mount, createLocalVue } from "@vue/test-utils"
import UsersList from "@/components/dashboard/UsersList"
import rotaUsersMock from "@/api/clavaAPI/rotasUsers/rotasResponse.mock"

describe("UsersList - loaded users", () => {
  let input
  let localVue

  beforeEach(() => {
    localVue = createLocalVue()

    input = {
      localVue,
      propsData: {
        users: rotaUsersMock.users
      }
    }
  })

  it("Snapshot test", () => {
    const wrapper = mount(UsersList, input)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it("Expect UsersList to emit selected users on clicking checkbox", () => {
    const wrapper = mount(UsersList, input)
    wrapper.findAll(".user-checkbox").at(0).find('[type="checkbox"]').trigger("click")

    expect(wrapper.emitted().change).toBeTruthy()
    expect(wrapper.emitted().change[0][0]).toEqual([1])
  })

  it("Expect UsersList to emit all selected users on clicking toggle get all shifts", () => {
    const wrapper = mount(UsersList, input)
    const toggleCheckBox = wrapper.find(".toggle-all")

    toggleCheckBox.trigger("click")

    expect(wrapper.emitted().change).toBeTruthy()
    expect(wrapper.emitted().change[0][0]).toEqual([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])

    toggleCheckBox.trigger("click")

    expect(wrapper.emitted().change).toBeTruthy()
    expect(wrapper.emitted().change[1][0]).toEqual([])
  })
})
