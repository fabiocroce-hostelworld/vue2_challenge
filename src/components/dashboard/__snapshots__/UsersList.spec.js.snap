// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersList - loaded users Snapshot test 1`] = `
<div class="v-card v-sheet theme--light">
  <div class="v-card__title"> Users </div>
  <div class="v-card__text">
    <div role="list" class="v-list v-sheet theme--light">
      <div role="listbox" class="v-item-group theme--light v-list-item-group">
        <div tabindex="0" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light">
          <div class="v-list-item__content">
            <div class="v-list-item__title">John Smith</div>
          </div>
          <div class="v-list-item__action">
            <div class="v-input user-checkbox theme--light v-input--selection-controls v-input--checkbox">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$checkboxOff</i><input aria-checked="false" id="input-6" role="checkbox" type="checkbox" value="">
                    <div class="v-input--selection-controls__ripple"></div>
                  </div>
                </div>
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div tabindex="0" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light">
          <div class="v-list-item__content">
            <div class="v-list-item__title">Ann Doe</div>
          </div>
          <div class="v-list-item__action">
            <div class="v-input user-checkbox theme--light v-input--selection-controls v-input--checkbox">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$checkboxOff</i><input aria-checked="false" id="input-11" role="checkbox" type="checkbox" value="">
                    <div class="v-input--selection-controls__ripple"></div>
                  </div>
                </div>
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div tabindex="0" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light">
          <div class="v-list-item__content">
            <div class="v-list-item__title">Nicole Woo</div>
          </div>
          <div class="v-list-item__action">
            <div class="v-input user-checkbox theme--light v-input--selection-controls v-input--checkbox">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$checkboxOff</i><input aria-checked="false" id="input-16" role="checkbox" type="checkbox" value="">
                    <div class="v-input--selection-controls__ripple"></div>
                  </div>
                </div>
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div tabindex="0" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light">
          <div class="v-list-item__content">
            <div class="v-list-item__title">Bo Derek</div>
          </div>
          <div class="v-list-item__action">
            <div class="v-input user-checkbox theme--light v-input--selection-controls v-input--checkbox">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$checkboxOff</i><input aria-checked="false" id="input-21" role="checkbox" type="checkbox" value="">
                    <div class="v-input--selection-controls__ripple"></div>
                  </div>
                </div>
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div tabindex="0" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light">
          <div class="v-list-item__content">
            <div class="v-list-item__title">Titan Cox</div>
          </div>
          <div class="v-list-item__action">
            <div class="v-input user-checkbox theme--light v-input--selection-controls v-input--checkbox">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$checkboxOff</i><input aria-checked="false" id="input-26" role="checkbox" type="checkbox" value="">
                    <div class="v-input--selection-controls__ripple"></div>
                  </div>
                </div>
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div tabindex="0" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light">
          <div class="v-list-item__content">
            <div class="v-list-item__title">Peter Pan</div>
          </div>
          <div class="v-list-item__action">
            <div class="v-input user-checkbox theme--light v-input--selection-controls v-input--checkbox">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$checkboxOff</i><input aria-checked="false" id="input-31" role="checkbox" type="checkbox" value="">
                    <div class="v-input--selection-controls__ripple"></div>
                  </div>
                </div>
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div tabindex="0" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light">
          <div class="v-list-item__content">
            <div class="v-list-item__title">Jake Smith</div>
          </div>
          <div class="v-list-item__action">
            <div class="v-input user-checkbox theme--light v-input--selection-controls v-input--checkbox">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$checkboxOff</i><input aria-checked="false" id="input-36" role="checkbox" type="checkbox" value="">
                    <div class="v-input--selection-controls__ripple"></div>
                  </div>
                </div>
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div tabindex="0" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light">
          <div class="v-list-item__content">
            <div class="v-list-item__title">Blake Dee</div>
          </div>
          <div class="v-list-item__action">
            <div class="v-input user-checkbox theme--light v-input--selection-controls v-input--checkbox">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$checkboxOff</i><input aria-checked="false" id="input-41" role="checkbox" type="checkbox" value="">
                    <div class="v-input--selection-controls__ripple"></div>
                  </div>
                </div>
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div tabindex="0" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light">
          <div class="v-list-item__content">
            <div class="v-list-item__title">Joana Li</div>
          </div>
          <div class="v-list-item__action">
            <div class="v-input user-checkbox theme--light v-input--selection-controls v-input--checkbox">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$checkboxOff</i><input aria-checked="false" id="input-46" role="checkbox" type="checkbox" value="">
                    <div class="v-input--selection-controls__ripple"></div>
                  </div>
                </div>
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div tabindex="0" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light">
          <div class="v-list-item__content">
            <div class="v-list-item__title">Beef Patty</div>
          </div>
          <div class="v-list-item__action">
            <div class="v-input user-checkbox theme--light v-input--selection-controls v-input--checkbox">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">$checkboxOff</i><input aria-checked="false" id="input-51" role="checkbox" type="checkbox" value="">
                    <div class="v-input--selection-controls__ripple"></div>
                  </div>
                </div>
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <hr role="separator" aria-orientation="horizontal" class="mb-4 v-divider theme--light"> <button type="button" class="toggle-all mr-4 primary v-btn v-btn--contained theme--light v-size--default"><span class="v-btn__content">
        Toggle all shifts
      </span></button>
    </div>
  </div>
</div>
`;
